Python/specialize.c

```diff
@@ -569,8 +577,47 @@ analyze_descriptor(PyTypeObject *type, PyObject *name, PyObject **descr, int sto
     }
     else {
         if (type->tp_getattro != PyObject_GenericGetAttr) {
-            *descr = NULL;
-            return GETSET_OVERRIDDEN;
+            getattrofunc getattro_slot = type->tp_getattro;
+            /* The following comment is copied from typeobject.c */
+            /* There are two slot dispatch functions for tp_getattro.
+
+               - _Py_slot_tp_getattro() is used when __getattribute__ is overridden
+                 but no __getattr__ hook is present;
+
+               - _Py_slot_tp_getattr_hook() is used when a __getattr__ hook is present.
+
+               The code in update_one_slot() always installs
+               _Py_slot_tp_getattr_hook(); this detects the absence of
+               __getattr__ and then installs the simpler slot if necessary. */
+            /* Using some custom tp_getattro, like super() objects */
+            if (getattro_slot != _Py_slot_tp_getattr_hook &&
+                getattro_slot != _Py_slot_tp_getattro) {
+                *descr = NULL;
+                return GETSET_OVERRIDDEN;
+            }
+            PyObject *getattribute = _PyType_Lookup(type,
+                &_Py_ID(__getattribute__));
+            has_custom_getattribute = getattribute != NULL &&
```

Comment:
This seems convoluted, and I'm not convinced it is correct.
Would this work?
```C
PyObject *object_getattribute = PyType_Lookup(&PyBaseObject_Type,&_Py_ID(__getattribute__));
has_custom_getattribute = getattribute != object_getattribute;
```
We should probably cache the value of `object_getattribute`

Reply:
I copied it from what `slot_tp_getattro` uses https://github.com/python/cpython/blob/main/Objects/typeobject.c#L7836-L7840.

So at the very least, it's consistent with how `typeobject.c` does it.

Reply:
After some thought, I'd prefer we keep it as-is. I'd rather follow the `typeobject.c` implementation closely than implement a more correct check which may deviate from how `typeobject.c` behaves.

Reply:
That check is in `slot_tp_getattr_hook`, which only applies if `__getattr__` is set. 
What if `__getattr__` is not set?

Reply:
Then it calls `__getattribute__` https://github.com/python/cpython/blob/main/Objects/typeobject.c#L7781-L7784. In this case since there's no `__getattr__` and no custom `__getattribute__`, it just fails through and specializes to normal attribute lookup. 
```pycon
>>> class X:
...  __getattribute__ = object.__getattribute__
...  def __init__(s):
...   s.x = 1
...
>>> x = X()
>>>
>>> def f():
...  for _ in range(10):
...   x.x
...
>>> f()
>>> import dis
>>> dis.dis(f, adaptive=True)
  1           0 RESUME_QUICK             0

  2           2 LOAD_GLOBAL_ADAPTIVE     1 (NULL + range)
             14 LOAD_CONST               1 (10)
             16 CALL_ADAPTIVE            1
             26 GET_ITER
        >>   28 FOR_ITER_RANGE          19 (to 70)
             32 STORE_FAST               0 (_)

  3          34 LOAD_GLOBAL_MODULE       2 (x)
             46 LOAD_ATTR_INSTANCE_VALUE     2 (x)
             66 POP_TOP
             68 JUMP_BACKWARD_QUICK     21 (to 28)

  2     >>   70 LOAD_CONST               0 (None)
             72 RETURN_VALUE
>>>
```

Reply:
You are copying code from `slot_tp_getattr_hook`, but that is only correct if `getattro_slot == _Py_slot_tp_getattr_hook` is set. 

Not only that, but there is no guarantee that class's version number changes when `type->tp_getattro` changes, so we cannot rely on the invariants implied by `slot_tp_getattr_hook`.

The `type->tp_getattro` dance is an implementation of the true semantics.
It is is independent of specialization and it may change in a way that is correct, but breaks your assumptions here.
For example, we may remove `PyWrapperDescr_Type` and `PyWrapperDescrObject` and replace them with builtin functions in future.

Reply:
> Not only that, but there is no guarantee that class's version number changes when type->tp_getattro changes, so we cannot rely on the invariants implied by slot_tp_getattr_hook

That's tough. You mean it could cycle between `_Py_slot_tp_getattr_hook` or `_Py_slot_tp_getattro` right? If it's just that we can check at runtime that things stay as expected.

If you mean that any arbitrary type could just modify its `type->tp_getattro` at runtime. That would mean all of our current `LOAD_ATTR` specialized opcodes will need runtime checking that `tp_getattro` stays as `PyObject_GenericGetAttr`. How do I know a C type doesn't change to a custom lookup function halfway through? However, the perf impact for such an unlikely case isn't worth it. I'd be willing to break such obscure code if it means slowing down Python overall by 1% just to support these.

> You are copying code from slot_tp_getattr_hook, but that is only correct if getattro_slot == _Py_slot_tp_getattr_hook is set.

I can add another check at specialization time that ``getattribute == object_getattribute`` like you suggested. But I don't see how the current (more permissive) check could be wrong either. If a wrapper descriptor wraps a `PyObject_GenericGetAttr`, how can its lookup be anything but standard?

Are you alright with me checking both, that is :
```C
            has_custom_getattribute = getattribute != NULL &&
                !((Py_IS_TYPE(getattribute, &PyWrapperDescr_Type) &&
                    ((PyWrapperDescrObject *)getattribute)->d_wrapped ==
                    (void *)PyObject_GenericGetAttr) ||
                    getattribute != object_getattribute);
```
?
When the time comes in the future when we change ``object.__getattribute__``. We can just change the specializing code.

Reply:
What I mean is that we shouldn't rely on how `_Py_slot_tp_getattr_hook` or `_Py_slot_tp_getattro` behave.
If either are set, then we should look up `__getattribute__` and `__getattr__` directly and use those.
All we should rely on is:
* `type->tp_getattro == PyObject_GenericGetAttr`: Normal behaviour.
* `type->tp_getattro == _Py_slot_tp_getattr_hook` **or** `type->tp_getattro == _Py_slot_tp_getattro`: The `tp_getattro` hasn't been specified in the type struct, but either `__getattribute__` or `__getattr__` may have been overridden.

The first case is fine. We already handle it.
The second case is a bit more complex, but well defined [here](https://docs.python.org/3/reference/datamodel.html#customizing-attribute-access). Rather than trying to decipher the code in typeobject.c, we should follow the docs.

Reply:
Would this work?
```C
bool has_getattr;
getattrofunc getattro_slot = type->tp_getattro;
if (getattro_slot == PyObject_GenericGetAttr) {
    /* Normal attribute lookup; */
   has_getattr = false;
}
else if (getattro_slot == _Py_slot_tp_getattro || getattro_slot == _Py_slot_tp_getattr_hook) {
    /* One or both of __getattribute__ or __getattr__ may have been overridden */ 
    PyObject *getattribute = _PyType_Lookup(type, &_Py_ID(__getattribute__));
    PyObject *getattr = _PyType_Lookup(type, &_Py_ID(__getattr__));
    bool has_custom_getattribute = getattribute != object_getattribute;
    has_getattr = getattr != NULL;
    if (has_custom_getattribute) {
        if (!has_getattr && Py_IS_TYPE(getattribute, &PyFunction_Type)) {
            *descr = getattribute;
            return GETATTRIBUTE_IS_PYTHON_FUNCTION;
        }
        *descr = NULL;
        return GETSET_OVERRIDDEN;
    }
}
else {
    /* Something else */
   *descr = NULL;
   return GETSET_OVERRIDDEN;
}
```

We could replace the logic for setting `has_getattr` with `has_getattr = getattro_slot == _Py_slot_tp_getattr_hook` which would be faster, but I'd like to add `assert(_PyType_Lookup(type, &_Py_ID(__getattr__)) == NULL)` in that case.

Reply:
Thanks for providing the code. This was really helpful.